lib/config/flat-config-schema.js

```diff
@@ -301,47 +312,46 @@ const deepObjectAssignSchema = {
     validate: "object"
 };
 
+
 //-----------------------------------------------------------------------------
 // High-Level Schemas
 //-----------------------------------------------------------------------------
 
 /** @type {ObjectPropertySchema} */
-const globalsSchema = {
-    merge: "assign",
-    validate(value) {
+const languageOptionsSchema = {
+    merge(first = {}, second = {}) {
 
-        assertIsObject(value);
+        const result = deepMerge(first, second);
 
-        for (const key of Object.keys(value)) {
+        for (const [key, value] of Object.entries(result)) {
 
-            // avoid hairy edge case
-            if (key === "__proto__") {
-                continue;
-            }
-
-            if (key !== key.trim()) {
-                throw new TypeError(`Global "${key}" has leading or trailing whitespace.`);
-            }
+            /*
+             * Special case: Because the `parser` property is an object, it should
+             * not be deep merged. Instead, it should be replaced if it exists in
+             * the second object. To make this more generic, we just check for
+             * objects with methods and replace them if they exist in the second
+             * object.
+             */
+            if (value && typeof value === "object") {
+                if (hasMethod(value)) {
+                    result[key] = second[key] ?? first[key];
+                    continue;
+                }
 
-            if (!globalVariablesValues.has(value[key])) {
-                throw new TypeError(`Key "${key}": Expected "readonly", "writable", or "off".`);
+                // for other objects, make sure we aren't reusing the same object
+                result[key] = deepMerge(first[key] ?? {}, second[key] ?? {});
```

Comment:
With this config file:

```
// eslint.config.js
module.exports = [{
    languageOptions: {
        globals: "foo"
    }
}];
```

When ESLint is run, CLI passes `globals: {}` (with global variables inside if `--global` flag was used, but it always passes at least an empty globals object) in `overrideConfig`.

This line calls `deepMerge('foo', {})`, which results in `{ '0': 'f', '1': 'o', '2': 'o' }`, and at the end produces the following cryptic error message:

```
TypeError: Key "languageOptions": "Key "globals": Key "0": Expected "readonly", "writable", or "off".
```

Maybe we could change this code to:

```js
if (isNonArrayObject(value)) {
    if (hasMethod(value)) {
        result[key] = second[key] ?? first[key];
        continue;
    }

    // for other objects, make sure we aren't reusing the same object
    result[key] = { ...result[key] };
}
```

That way, the invalid `globals: "foo"` would be silently ignored, but that's a drawback this generic merging already introduces because it can't validate each config object.

Reply:
Good catch. Updated.